--- a/sklearn/_build_utils/openmp_helpers.py
+++ b/sklearn/_build_utils/openmp_helpers.py
@@ -40,6 +40,8 @@
         # export LDFLAGS="$LDFLAGS -Wl,-rpath,/usr/local/opt/libomp/lib
         #                          -L/usr/local/opt/libomp/lib -lomp"
         return []
+    elif sys.platform == "linux" and 'musl' in compiler:
+        return []
     # Default flag for GCC and clang:
     return ['-fopenmp']
 
